### ISSUE TITLE
invisible delimiter in macro expansion

### ISSUE THREAD

**Issue by edwin0cheng** [rust-analyzer/rust-analyzer#1132]
As discussion in PR #1125 , we want to add 2 two new token kinds (`L_DOLLAR` and `R_DOLLAR`) for fixing the parser such that it doesn't break `Delimiter::None` subtrees. 

### Background
However, i just find out the current parser did not break by `Delimiter::None` in simple way. For example:

*rules*
```rust
        macro_rules! foo {
            ($ i:expr) => {
                 fn bar() { $ i * 10; } 
            }
        }   
```
*invocation*
```rust
"foo! { 2 + 2 }",
```
It will expand to `fn bar () {2 + 2 * 10 ;}` , which seem to be wrong, because we would like to get back `fn bar() {(2+2) * 10; }` for preserving the execution order.

In the `ast` tree point of view, the macro expansion of `(2+2) * 10` would look like this:
```rust
BIN_EXPR
    LITERAL
        INT_NUMBER
    PLUS
    BIN_EXPR
        LITERAL
            INT_NUMBER
        STAR
        LITERAL
            INT_NUMBER
```

The suggested solution in the previous discussion is adding two token kinds: `L_DOLLAR` and `R_DOLLOR` for invisible delimiter in macro expansion to fix this problem.

### Proposal 

But after i do some research about how `rustc` mbe works, i will proposal add a new `MARCO_INTERPOLATED` node *after* macro expansion. so the `ast` tree would became:
```rust
BIN_EXPR
    MARCO_INTERPOLATED
        LITERAL
            INT_NUMBER
        PLUS
        LITERAL
            INT_NUMBER
    STAR
    LITERAL
        INT_NUMBER
```

The reasons are:
1. It simplify the parser logic: We don't need to add new token kinds.
2. Even we add the new token kinds, we still need to add a new node for contain it:
```rust
BIN_EXPR
    DOLLOR_EXPR    <- Dolar expression ??
        LITERAL
            INT_NUMBER
        PLUS
        LITERAL
            INT_NUMBER
    STAR
    LITERAL
        INT_NUMBER
```
3.  Right now `Delimiter::None` only happens in the top level of macro expansion and it should not be seen in later stage of parsing.

However, i am not sure how it affect `hir` code and other parts of `ra`.

**Comment by matklad** [comment]
An excellent observation! My understanding is that interpolated tokens in `rustc` are specific for macro-by-example, and are not available for proc-macros. In rust-analyzer, I would love to handle proc-macros and mbe macros using the exact same API of token trees (this is why tt crate exists separately from the rest of the compiler), and proc_macro's tt do not have interpolated nodes.

If we don't touch `ra_tt`, we still can add interpolated nodes to mbe specifically, by mirroring `ra_tt` trees inside mbe (which we do anyway for `$captures`).

However, I think we must handle `None`s in parer correctly anyway, because proc-macros would use that. in other words, I think that if `foo` is written as a proc_macro, it still gets the correct expansion in rustc, without using interpolated tokens (i might be wrong here). 


So, I think the way forward here is

1. verify the theory that we should handle `None` anyway for proc macros
2. implement mbe using `None`
3. (future) improve mbe by using interpolated: this should be more efficient, b/c we will be able to share green trees. Moreover, I think *correct* handlng of mbe requires interpolated, because mbe forbid to capture and `$e:expr` and then reparsing it as `$ty:ty`, but the `Delimiter::None` approach allows it.

**Comment by matklad** [comment]
>i will proposal add a new MARCO_INTERPOLATED node after macro expansion. so the ast tree would became:

I don't think we need `MACRO_INTERPOLATED` in the final tree, it should look like

```
BIN_EXPR
    BIN_EXPR
        LITERAL
            INT_NUMBER
        PLUS
        LITERAL
            INT_NUMBER
    STAR
    LITERAL
        INT_NUMBER
```

That is, if you print this tree, you get a "wrong" result, but we care about the tree structure, and it could be correct even without introducing an additional node

**Comment by matklad** [comment]
@edwin0cheng if you are around, we have an MBE meeting in the Zulip right now ;)

**Comment by edwin0cheng** [comment]
@matklad Just want to clarify, if i expand following marco :
```
macro_rules! structs {
    ($($i:ident),*) => {
        $(struct $i { field: u32 } )*
    }
}
```

 Is it the syntax tree output with L_DOLLAR and R_DOLLAR you expected ?
```
SOURCE_FILE@[0; 40)
  L_DOLLAR@[0; 0) ""
  L_DOLLAR@[0; 0) ""
  L_DOLLAR@[0; 0) ""
  STRUCT_DEF@[0; 20)
    STRUCT_KW@[0; 6) "struct"
    NAME@[6; 9)
      IDENT@[6; 9) "Foo"
    NAMED_FIELD_DEF_LIST@[9; 20)
      L_CURLY@[9; 10) "{"
      NAMED_FIELD_DEF@[10; 19)
        NAME@[10; 15)
          IDENT@[10; 15) "field"
        COLON@[15; 16) ":"
        PATH_TYPE@[16; 19)
          PATH@[16; 19)
            PATH_SEGMENT@[16; 19)
              NAME_REF@[16; 19)
                IDENT@[16; 19) "u32"
      R_CURLY@[19; 20) "}"
  R_DOLLAR@[20; 20) ""
  L_DOLLAR@[20; 20) ""
  STRUCT_DEF@[20; 40)
    STRUCT_KW@[20; 26) "struct"
    NAME@[26; 29)
      IDENT@[26; 29) "Bar"
    NAMED_FIELD_DEF_LIST@[29; 40)
      L_CURLY@[29; 30) "{"
      NAMED_FIELD_DEF@[30; 39)
        NAME@[30; 35)
          IDENT@[30; 35) "field"
        COLON@[35; 36) ":"
        PATH_TYPE@[36; 39)
          PATH@[36; 39)
            PATH_SEGMENT@[36; 39)
              NAME_REF@[36; 39)
                IDENT@[36; 39) "u32"
      R_CURLY@[39; 40) "}"
  R_DOLLAR@[40; 40) ""
  R_DOLLAR@[40; 40) ""
  R_DOLLAR@[40; 40) ""
```

PS: The first `Delimiter::None` is added in parsing the whole macro (rhs), the second one is parsing the `$repeat`, the third one is expanding the `$repeat`

**Comment by matklad** [comment]
@edwin0cheng no, I would expect that there wouldn't be any explicit dollars in the final tree. Basically, when the tree-building TreeSink sees a dollar token, it should just discard it. The nesting structure of the tree should be correct nonetheless. 

In other words, dollars are for parser, but not for the final tree

**Comment by edwin0cheng** [comment]
Ah.....But current parser cannot discard a token...

**Comment by matklad** [comment]
Yeah, it's the tree sink who will discard the token, not the parser

**Comment by matklad** [comment]
SPecifically, here shold be an `if kind != L_DOLLAR`: https://github.com/rust-analyzer/rust-analyzer/blob/e6e2571bdf780d304c792d4317bbaf1d6f5d7a0a/crates/ra_mbe/src/syntax_bridge.rs#L137

**Comment by edwin0cheng** [comment]
It fixed in #1138
